src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt

```diff
@@ -825,6 +826,7 @@ class BookingService(
 
     booking.arrivalDate = arrivalDate
     booking.departureDate = expectedDepartureDate
+    booking.status = BookingStatus.arrived
```

Comment:
Setting this after the ArrivalEntity is created makes sense because if that fails, we don't want to set the status. The way this is order dependent feels a touch fragile and probably applies to the other state changes too.

I'm wondering if an after create hook (if that's even possible) could be set up on the arrival entity so that if this line gets moved around the two states can't be out of sync. Whenever an arrival entity is created that entity goes and updates the parent booking, hopefully this would provide a better guarantee. That would mean even more UPDATEs though which isn't great either.

What are everyone's thoughts?

Reply:
We do a similar thing for applications using Listeners (see https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1119)

Reply:
I would prefer the service owns the responsibility of updating the booking status , so that the service has control over when and how a booking status get updated.
i'm not sure even we can add EventListener for this as such we are updating different entities.

Reply:
I've just spotted that these methods are marked with `@Transactional` so if the arrival fails to create then we should never have the booking.status=arrived. 

So long as we have this on all the service methods I think we are protected from someone in the future unknowingly changing the ordering of these calls. 👍 

On the subject of listeners I think it comes down to appetite to risk. We should be able to trust our services to take this responsibility however there is always a risk that in future one of the many services are changed unknowingly, a new service method takes over or even records start being created outside of services(!). Using listeners does, I think, provide a better guarantee of the status always being updated. 

Perhaps we go with this and if it becomes flakey we pivot to listeners?

I hope we could [write a post-create action that fetches the parent booking and updates it](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1119/files#diff-3554486b057c692da1ac1f3b75addc76f71670c8efbdc4594e42eda5b8edfed8R11).